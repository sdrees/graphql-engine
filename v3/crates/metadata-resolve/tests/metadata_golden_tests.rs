//! Tests that attempt to resolve different metadata files and assert that they parse successfully
//! or fail in the expected way.

use metadata_resolve::MetadataResolveFlagsInternal;

#[test]
fn test_passing_metadata() {
    insta::glob!("passing/**/metadata.json", |path| {
        insta::with_settings!({
            snapshot_path => path.parent().unwrap(),
            snapshot_suffix => "",
            prepend_module_to_snapshot => false,
        }, {
            let metadata_resolve_flags_internal = MetadataResolveFlagsInternal {
                enable_boolean_expression_types: true,
            };

            let metadata_json_text = std::fs::read_to_string(path)
                .unwrap_or_else(|error| panic!("Could not read file {path:?}: {error}"));

            let metadata_json_value = serde_json::from_str(&metadata_json_text)
                .unwrap_or_else(|error| panic!("Could not parse JSON: {error}"));

            let metadata = open_dds::traits::OpenDd::deserialize(metadata_json_value)
                .unwrap_or_else(|error| panic!("Could not deserialize metadata: {error}"));

            let resolved = metadata_resolve::resolve(metadata, metadata_resolve_flags_internal)
                .unwrap_or_else(|error| panic!("Could not resolve metadata: {error}"));

            insta::assert_debug_snapshot!("resolved", resolved);
        });
    });
}

#[test]
fn test_failing_metadata() {
    insta::glob!("failing/**/metadata.json", |path| {
        insta::with_settings!({
            snapshot_path => path.parent().unwrap(),
            snapshot_suffix => "",
            prepend_module_to_snapshot => false,
        }, {
            let metadata_resolve_flags_internal = MetadataResolveFlagsInternal {
                enable_boolean_expression_types: true,
            };

            let metadata_json_text = std::fs::read_to_string(path)
                .unwrap_or_else(|error| panic!("Could not read file {path:?}: {error}"));

            match serde_json::from_str(&metadata_json_text) {
                Ok(metadata_json_value) => {
                    match open_dds::traits::OpenDd::deserialize(metadata_json_value) {
                        Ok(metadata) => {
                            match metadata_resolve::resolve(metadata, metadata_resolve_flags_internal) {
                                Ok(_) => {
                                    panic!("Unexpected success when resolving {path:?}.");
                                }
                                Err(msg) => {
                                    insta::assert_snapshot!("resolve_error", msg);
                                }
                            }
                        }
                        Err(msg) => {
                            insta::assert_snapshot!("deserialize_error", msg);
                        }
                    };
                }

                Err(msg) => {
                    insta::assert_snapshot!("parse_error", msg);
                }
            };
        });
    });
}
