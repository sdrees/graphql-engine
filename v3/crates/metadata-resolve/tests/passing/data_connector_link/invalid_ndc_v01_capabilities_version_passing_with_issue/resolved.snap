---
source: crates/metadata-resolve/tests/metadata_golden_tests.rs
expression: resolved
input_file: crates/metadata-resolve/tests/passing/data_connector_link/invalid_ndc_v01_capabilities_version_passing_with_issue/metadata.json
---
(
    Metadata {
        object_types: {},
        scalar_types: {},
        models: {},
        commands: {},
        object_boolean_expression_types: {},
        boolean_expression_types: BooleanExpressionTypes {
            objects: {},
            scalars: {},
        },
        aggregate_expressions: {},
        graphql_config: GlobalGraphqlConfig {
            query_root_type_name: TypeName(
                Name(
                    "Query",
                ),
            ),
            mutation_root_type_name: TypeName(
                Name(
                    "Mutation",
                ),
            ),
            order_by_input: Some(
                OrderByInputGraphqlConfig {
                    asc_direction_field_value: Name(
                        "Asc",
                    ),
                    desc_direction_field_value: Name(
                        "Desc",
                    ),
                    enum_type_name: TypeName(
                        Name(
                            "order_by",
                        ),
                    ),
                },
            ),
            enable_apollo_federation_fields: false,
        },
        roles: [],
    },
    [
        DataConnectorIssue(
            NamedDataConnectorIssue {
                data_connector_name: Qualified {
                    subgraph: "default",
                    name: DataConnectorName(
                        Identifier(
                            "myconnector",
                        ),
                    ),
                },
                issue: InvalidNdcV01Version {
                    version: "^0.1.0",
                    error: Error("unexpected character '^' while parsing major version number"),
                },
            },
        ),
    ],
)
